### ISSUE TITLE
event_multilingual: Issues with 2.3

### ISSUE THREAD

**Issue by th-h** [s9y/Serendipity#635]
When changing the display language, everything works as designed at first ... until you click on a link. Afterwards, the entry is shown in the default language, but some other elements (like navigation) are translated, while others are not. See the [forum thread](https://board.s9y.org/viewtopic.php?f=4&p=10452371#p10452371).

This seems due to changes in `functions_config.inc.php`. See the [forum thread](https://board.s9y.org/viewtopic.php?f=3&t=24544#p10452379).

(Didn't try to reproduce myself due to lack of time, just relaying from the forums here.)

**Comment by stephanbrunker** [comment]
After I finally managed to copy my backup of the 2.1.2 installation to my development server, i have some additional information:

The problem is not related to the 2.3 update because also 2.1.2 shows the problem. If on the page displayed none of the articles has a translation to the selected language, then it jumps partially back to the default language. I am trying to look into the issue now by copying the 2.3 also to the development server and then I can start to debug the issue with var dumps into the log or I am at least going to try to fix it. And if I am in the mood, also make the archive plugin multilingual, showing only the entries in the selected language. And/or adding an option to show only the entries which have a translation. Let's see.

**Comment by stephanbrunker** [comment]
debugging the issue: it is quite complicated because there are so many variables associated with the language. In the situation above, GetSessionLanguage returns correctly 'en', but at the end of serendipity_config.inc.php, $serendipity['lang'] is 'de' while $serendipity['COOKIE'] is still 'en' and $_SESSION['serendipityLanguage'] is also 'de'. I have to look into $serendipity['autolang'] yet. What a mess … normally, I'd say that the one who did this has to fix it …

**Comment by garvinhicking** [comment]
;)

It worked at some point, maybe jumping back to a s9y 2.0 or 1.9 version can easier show what changed that broke it?

Your description may suggest that the s9y setcookie method might‘ve been changed and doesnt directly change the cookie value?

Also, maybe your browser language autodetection could be enabled which might take precedence. Or you are logged in, in which case your author‘s languge takes precedence... did you try it running a browser where you are not logged into s9y?

Hth,
Garvin

**Comment by stephanbrunker** [comment]
atm, I am working my way though the code. I don't think it is setcookie, that was my first idea because that changed from 2.1.2 to 2.3, but that was not true because the problem also existed back then. 

I am not logged in, somewhere between line 340 and the end of serendipity_config.inc.php the value of 'lang' changes without be called directly. It can be the autodetect which is called thrice: once via 'include lang.inc.php' in line 195, once via GetSessionLanguage in line 340 and then again with 'lang.inc.php' in line 396.

**Comment by stephanbrunker** [comment]
No, it is not the autodetect. I have to do it the hard way and log the value of 'lang' at various points between 340 and the end to see where it is changed ...

**Comment by ophian** [comment]
Just to chim in. The **key point** is to use a different browser for frontend testing! And better have global and personal lang set the same. (This both, while there are so many associate game changers.)
See https://github.com/ophian/additional_plugins/blob/master/serendipity_event_multilingual/documentation_de.html#L69 ff.
I can attest that [Serendipity Styx](https://github.com/ophian/) version of multilingual plugin is working well.

**Comment by stephanbrunker** [comment]
The issue is that it is not working with Edge in German setting, so that has to be fixed anyway, even if it works in other combinations. If the 3rd gear of your car is gone, you can still drive but that is not the way it should be. I'm just stuck because I cannot find an entry point for this call:

serendipity_plugin_api::hook_event('backend_login', $is_authenticated, NULL);

a full-text keyword search for 'backend_login' returns nothing, just hits in the template. But there has to be an associated function because before this call the $serendipity['lang'] variable is 'en' and after it is the browser (or default) language 'de'. And as far as I could determine of the logic so far, the language should be set at this point, when almost everything is already loaded.

**Comment by ophian** [comment]
>The issue is that it is not working with Edge in German setting, so that has to be fixed anyway, even if it works in other combinations.

What does that mean exactly?
Give me a tight example please.

Did you read all related notes in pasted documentary? Yes, it is very difficult to point this out clearly debugging-wise, since you have Cookies, Sessions, Lang constants, Serendipity global array variables, GET variables and all sort of Serendipity and Plugin options **and** your current browser settings interacting with each other, all differentially prioritized. I checked my testing blogs with certain firefox and chrome browsers and all worked fine. You have to make sure testing visitors behaviors, with a non-spoiled browser!

**Comment by stephanbrunker** [comment]
As I wrote above, I have Edge on a german Windows 10. If I change the language with the drop-down field to english, the POST variable is set, also the cookie and 'lang' variable. If I click then on the bottom of the page on the 'next' or 'previous' link, everything is okay, the serendipity_config.inc.php runs through the first 340 lines, the getSessionLanguage returns 'en', cookie is set et cetera. All correct. But then, in the call to serendipity_login(), inside that serendipity_authenticate_author() and inside that the hook_event 'backend_login'. And exactly after this one specific call, the 'lang' variable switches back to 'de'. The result is that everything which is dependent on that 'lang' variable is in german, and what is dependent on the cookie or loaded before that call is in english, like the variables PLUGIN_STATICPAGELIST_FRONTPAGE_LINKNAME or PLUGIN_SIDEBAR_MULTILINGUAL_TITLE. That is pretty simple. For further debugging, I just need to know what happens on the event_hook call.

**Comment by ophian** [comment]
Presumably because it is the "non-saved" "temporary" (lang) solution.
Please read and check:
https://github.com/ophian/additional_plugins/blob/master/serendipity_event_multilingual/documentation_de.html#L199

**Comment by stephanbrunker** [comment]
I  have read and checked, it is a known behavior. But even if there is a workaround in the options, the question is: is it possible to fix it at the source? And also in the FAQ is the issue that it is not (simply) possible to hide the non-translated articles in a second language. But you never know until you have tried, the same is true for the archive plugin. Sounds like I have to dig deep into the inner workings of serendipity, understand how it works and then working on a solution. Some people solve Sodokus for brain activity, I'm going for stuff like that ...

**Comment by ophian** [comment]
Its not a bug or weird behaviour to "fix". It is a **feature**.
And the multilingual option is not a "workaround", than a **feature set**. :)

I just added the following to the FAQ N° 3 question
https://github.com/ophian/additional_plugins/blob/master/serendipity_event_multilingual/documentation_de.html#L210
which should answer your question about "hiding" default language entries.
This is a documented behaviour in the Serendipity Book:
https://ophian.github.io/book/#serendipity-event-multilingual

The archive plugin shows the entries in the selected language, and all others since they are the fallback (normal) entries. The customarchive did miss to check for multilingual translated titles. This is fixed in Styx.

**Comment by stephanbrunker** [comment]
Sorry, but the serendipity_config.inc.php file looks very much that it is intended that the cookie value and the $serendipity['lang'] value are the same. There is even a function for that and certainly who wrote line 1049 of functions_config.inc.php:

$_SESSION['serendipityLanguage'] = $lang;

intended it to be the same.  And when it is suddenly not the same and the blog shows two languages at the same time, that seems slightly wrong, please correct me. I offer the non-recurring offer that I look personally in depth into that issue and invest my time which is very valuable for me. Because the product doesn't behave logically which is how software should work and I also think that an option to hide non-translated articles would be a very valuable extension, too. I did a few fixes and extensions in the past and if you don't want my contribution, I simply don't have to.

**Comment by th-h** [comment]
>  I offer the non-recurring offer that I look personally in depth into that issue and invest my time which is very valuable for me. Because the product doesn't behave logically which is how software should work and I also think that an option to hide non-translated articles would be a very valuable extension, too. Just think about the situation if the most recent articles aren't translated yet and the visitor doesn't see anything in his language on the first impression. I did a few fixes and extensions in the past and if you don't want my contribution, I simply don't have to.

We really do like each and every contribution to Serendipity, and I would very much appreciate it if you could spend the time to get to the bottom of it. The issue - or part thereof - may be fixed in Styx (which is @ophian's fork of Serendipity), but it doesn't look right to me in Serendipity proper.

**Comment by ophian** [comment]
Sounds somehow pissed-off and leaves me a little perplex, since I don't know why...
This is not my plugin!
I did some fixes and improvements in the past, playing "Sudoku", trying to keep it alive and did some more in the last years for Styx. I had to dive deep into to get a little bit of what was originally intended. If you proof these fixes or the origin intend nuts, please go on and give a better solution. I just wanted to give an idea of what is associated and meant before you try to fix issues which are meant to be features, (_or aren't-ready-yet-features, which is a S9y classical in the last 4 years_). Serendipity is well designed and these extended features need some care and love at all. It does not say, it does not have bugs though.
And you have to keep in mind that Serendipity is not a CMS designed to use multilingual natively. The multilingual plugin just adds some functionality into the origin design. To fully do multilangs, better have two independent blogs or use categories!

The "issue" showing 2 languages at the same time (the default and the user-selected) is something that is intended,  IMHO, since you have a weblog in a default language. Multilang entries are an **addition**, nothing else, and in special **not** a replacement for the first. The last mentioned GET variable option just makes it possible to show up certain languages only. So a foreign visitor visits a german blog. You give him the opportunity to read certain entries in his native language. This is what the entries/archives lists says: German blog with german entries, some of them translated for your pleasure. You even do not need the sidebar selection. You can give a foreigner some navigation or nugget links to view translated entries only. And you can declare this extra somewhere. Sidebar selection does even more and uses translated Serendipity or Plugin structural constants if there is a need for it.

**Comment by garvinhicking** [comment]
What is even going on here? To me this actually reads like you, Ian, are offending people. 

Please keep this civil and productive, I do think stephan is really trying to get to the bottom of things and the multilingual stuff is really really messy, so every hand helps.

**Comment by ophian** [comment]
@garvinhicking 
Hmm.., I do not see any offending note in my tries to give a helpful answer, beside that "_friendly s9y sidekick_" in the last post. If that is offending you, I am sorry.

Well Garvin, you as the origin author of this plugin know best what was intended and which historic reasons led to this (...). Please give advise.

**Comment by stephanbrunker** [comment]
I think my original post in the forum was not read carefully and because of that, someone jumped to conclusions way too early. Which is kind of funny because part of my blog is dedicated to that. So, to illustrate the bug, i made a screenshot what happens on a 'de' blog with a 'de' browser after you switch the language to 'en' and then click on 'next page':

![s9ybug](https://user-images.githubusercontent.com/18150071/65504364-1002fc00-dec7-11e9-95be-b5d674a3e1fb.png)

If that is a feature that seems to be a strange one. Only the title of the multilingual sidebar plugin and the title of the start page link in the staticpage sidebar plugin remain in english, everything else jumps back to 'de'. And the only way to reset that is to cycle the languages fully around. I posted that information from the beginning, so everyone commenting should have read it. 

Of course I know of the option 'force full language switch' and that makes sense: if you click on the language version link of an article, only that article or the whole blog changes language and you can decide which for yourself. And I'd add another option to hide non-translated articles if the blog language is changed away from default, if I can make that happen. And then working on the archive plugin too, which doesn't seem translated at all. Put perhaps I am going to fail because it is really impossible to do that, but that is my risk.

But I simply really, really don't like it if someone states that something I noted as clearly not intended is a feature and I should keep my hands off: _'it is a feature, not a bug'_ because that means that even if I do fix it for myself, it doesn't get implemented in the major fork and then the benefits of my work are reduced a lot.

**Comment by stephanbrunker** [comment]
Oh, I did read the programmers' documentation and together with that information I chased the $serendipity['lang'] variable through the complete startup process. It is actually the multilingual plugin changing the value when the first call to the event_hook api tries to get the plugin array, finds it empty and loads the plugins, Then the introspect function of the multilingual plugin makes some kind of initialization and changes the 'lang' variable back to 'de' but not the cookie and other lang-based variables. But that seems only to be the first part of the problem.

**Comment by ophian** [comment]
> But I simply really, really don't like it if someone states that something I noted as clearly not intended is a feature and I should keep my hands off: 'it is a feature, not a bug' because that means that even if I do fix it for myself, it doesn't get implemented in the major fork and then the benefits of my work are reduced a lot.

"someone" seems to be me then ;-) The rest is a misinterpretation.
I see your problem and I can see it happen on you blog, now that you have given me the red drawed circles. I assume you still have 'force full language switch' set to **no**.
This is where we talked at cross purposes and my extended plugin is probably making some certain differences, since on my local test blog I don't have that phenomenon. This is what I proofed to happen working "as designed".
1. thing I did - kill the cookie to start from scratch.
2. thing - drove to the page with a known multilingual entry.
3. checked everything was displayed in german since global and personal admin was set to German, then changed the sidebar selection to English.
4. Result: The same page brought in the en translated entry, sidebars quicksearch title turned into english, checked Cookie being set to serendipity[serendipityLanguage] = en.
5. Went down to paginate to "next "page: Checked quicksearch title was in German, checked archives month were in German etc.. Checked Cookie was still set to en. Went back with (german) paginate "vorherige Seite", everthing was in geman too. So all was fine.
That is the difference and **knowing** that the **cookie set lang** only plays its role when it is the single request OR is used permanently by  'force full language switch' true, if not it is ignored by plugin further on AFAIR.

So you could make a diff between Styx and S9y current versions of the multilingual plugin and see what might be the differing case.

There are so many possibilities with this plugin and I personally see it more as an extended addition.
I would not even say I have fully got and checked all these possibilities interacting with each other without flaws. I have some text cases and thats it. Examplewise a german entry copy-translated into english, an english entry without german origins, various {{lang}} blocks (mostly titles), etc.

**Comment by stephanbrunker** [comment]
I just see - it was always in front of my nose - that the issue is known and at 'TODO' status:

>  * This function also sets HTTP Headers and cookies to contain the language for follow-up requests
>  * TODO:
>  * This previously was handled inside a plugin with an event hook, but caching
>  * the event plugins that early in sequence created trouble with plugins not
>  * having loaded the right language.
>  * Find a way to let plugins hook into that sequence :-)
>  *
>  * @access public
>  * @return  string      Returns the name of the selected language.
>  */
> function serendipity_getSessionLanguage()

atm, it looks very much that this function and the initialization of the multilingual plugin are working against each other because that also does a lot of the stuff that function does. I am trying to sort that out by creating the rules of precedence for the language selection. Whoever wrote that TODO, I agree that setting the language more or less randomly at the first event_hook call was a bad idea, because you don't know in which order the plugins are initialized. Unfortunately, if the GET 'lang_selected' changes the language of the blog or not is dependent on the setting in the plugin which is still fetched later. I have to sort that out first. But on the other side, you can control the order of the plugins, you just have to document that the multilingual_event has to be the first one.

**Comment by stephanbrunker** [comment]
Interesting … as I am working through the plugin, I have the same ideas the developers had. You can solve that problem above by generating two different GET requests. And i found them just now, what has not been done is to implement GET 'language_selected' and GET 'user_language' correct into getSessionLanguage, this is wrong. If I go after the plugin, 'language_selected' doesn't change the blog language and 'user_language' does. Then I can remove everything what changes COOKIE, SESSION or the serendipity['lang'] variable from the plugin and it should be fine. The getSessionLanguage then controls the blog language and the plugin controls the entry language with $this->showlang and $this->lang_display. And if I read the comments in the code, $this->lang_display should display nothing if the language does not exist.

**Comment by ophian** [comment]
Before you go and add possible TODOs, may I ask if you have checked wth Styx multilingual v.2.61 first?
Did you problem resist in this case?

**Comment by stephanbrunker** [comment]
The TODO was already there, I didn't add it. I see the need to process it because I like to solve problems at their root and not the symptoms. The root of the problem is that both the getSessionLanguage in config.inc.php and the introspect function in multilingual_event.php try to change the blog language and that just calls for problems in both s9y and Styx fork. These two functions are the first step and I am going to create and document the language selection new from the beginning. For that I can use the existing blocks of code and variables, just sorting the order out and clean it up so that it remains fully compatible. I am going to comment it heavily and put it on my repo first, then you can decide what to do out of it. The only thing I am going to change is a new option for hiding entries when they don't exist in the requested language or displaying them in the default language. But this functionality is already prepared in the plugin with the two variables $this->showlang and $this->lang_display. According to the comments, using the latter should hide the entry:

>         // Okay, Garv. I explain this to you ONCE and ONLY.
>         // $this->lang_display is the variable that FORCES translations of entries. If a translation does not exist,
>         //                     an entry is NOT SHOWN.
>         // $this->showlang     is the variable that indicates which language of an entry to prefer.

**Comment by ophian** [comment]
Yes I do understand this, and I don't want to stop you trying to solve issues at their root.
But..., this function is very old. We have no proofed history of what this note was meant for, except Garvins mind remembering clearly the issue and what was related to it. We can't even say the multilingual plugin was the only and rooting issue for it, since all the early days commit histories are lost. What we know is that this function has proofed its stability at least since 2006. So touching this central core will led to a much deeper check job (including all possible plugins) than a special fix for multilingual and its relations will do, I fear. This is why I asked if you had tried with 2.61 first and still experienced the issue. Checking for a solution in the plugin first seems a good place to me, since that is the place where you have found issues and the relation is build or not.

**Comment by stephanbrunker** [comment]
I don't think my changes would cause additional problems. I'll keep the structure and the variables and what I did from the start was chasing the variables through the repository. And these are the only two places so far where these variables are written to, so touching the code here should be okay. I made a complete dump of the touched vars at the end of serendipity_config_inc.php and that should be consistent. I do code for fun and for me it is also a kind of art to make it as simple, clear and structured as it can be and in my own projects when I got a piece of code running I didn't stop there and also cleaned it up to get the same result with simpler code when possible. So when I have to work with it again a few years later I have a change of understanding what I did at the time.

Simply the fact that $_REQUEST, $_GET and $serendipity['GET'] etc. are touched simultaneously is everything else but not clean because these are multiple copies of the same information but only some are changed and others not. If that doesn't cause confusion, I don't know what does. The PHP vars are copied to the $serendipity array once and from there, only this copy should be used. And so far, only the multilingual plugin touches the PHP original vars. 

But I'll see.

**Comment by onli** [comment]
@stephanbrunker Feel definitely invited to look into improving this. We'll never have too many developers optimizing ;) 

Specific questions towards this best in separate issues, PRs or contact one of us directly.

**Comment by stephanbrunker** [comment]
As I now have this issue fixed in my develop version, i can call this one here closed. Now I got everything in the frontend to work as it should, I even could activate the option that entries are hidden if no translation is available. Same with the enforced language switch or not, that is okay so far. I have now less code and less variables and more comments how it works. 

Next steps are getting the plugins like recententries or archive fully multilingual and then the backend. I think it gets interesting if I have two users with different languages and then determine what is the 'default' language. At least I can think I can fix the backend display language to the language setting of the user in the database.

**Comment by fe-hicking** [comment]
Hey Stephan,

sorry for not being able to contribute at this point. Your work is huge, all this workflow has been a real pain, and I fully agree for it not being very clean. I applaud the work on cleaning it up and giving it a proper structure.

I am all for merging your changes into the plugin, I think we can only make progress for there, and if people have issues with the new plugin version, we can probably better address them with your updated code.

Regards,

Garvin

**Comment by stephanbrunker** [comment]
Uhh … done. I just did two pull requests in the Serendipity and additional_plugins repo which should fix all the issues I found. I just cannot guarantee that there aren't some side-effects with stuff I don't have installed. But so far, everything works now as intended in my blog and now I can pass around the link to the english version (default is DE) and the visitor doesn't see a lot of untranslated articles in a foreign language first.
